sdk/data_kusto/src/connection_string.rs

```diff
@@ -0,0 +1,336 @@
+            && self.authority_id == other.authority_id
+            && self.msi_auth == other.msi_auth
+            && self.msi_params == other.msi_params
+            && self.az_cli == other.az_cli
+    }
+}
+
+impl<'a> ConnectionString<'a> {
+    pub fn new(connection_string: &'a str) -> Result<Self, ConnectionStringError> {
+        let mut data_source = None;
+        let mut federated_security = None;
+        let mut user_id = None;
+        let mut user_token = None;
+        let mut password = None;
+        let mut application_client_id = None;
+        let mut application_token = None;
+        let mut application_key = None;
+        let mut application_certificate = None;
+        let mut application_certificate_thumbprint = None;
+        let mut authority_id = None;
+        let mut msi_auth = None;
+        let mut msi_params = None;
+        let mut az_cli = None;
+
+        let kv_str_pairs = connection_string
+            .split(';')
+            .filter(|s| !s.chars().all(char::is_whitespace));
+
+        for kv_pair_str in kv_str_pairs {
+            let mut kv = kv_pair_str.trim().split('=');
+            let k = match kv.next() {
+                Some(k) if k.chars().all(char::is_whitespace) => {
+                    return Err(ConnectionStringError::ParsingError {
+                        msg: "No key found".to_owned(),
+                    })
+                }
+                None => {
+                    return Err(ConnectionStringError::ParsingError {
+                        msg: "No key found".to_owned(),
+                    })
+                }
+                Some(k) => k,
+            };
+            let v = match kv.next() {
+                Some(v) if v.chars().all(char::is_whitespace) => {
+                    return Err(ConnectionStringError::MissingValue { key: k.to_owned() })
+                }
+                None => return Err(ConnectionStringError::MissingValue { key: k.to_owned() }),
+                Some(v) => v,
+            };
+
+            match k {
+                DATA_SOURCE_NAME => data_source = Some(v),
+                USER_ID_NAME => user_id = Some(v),
+                USER_TOKEN_NAME => user_token = Some(v),
+                PASSWORD_NAME => password = Some(v),
+                APPLICATION_CLIENT_ID_NAME => application_client_id = Some(v),
+                APPLICATION_KEY_NAME => application_key = Some(v),
+                APPLICATION_TOKEN_NAME => application_token = Some(v),
+                APPLICATION_CERTIFICATE_NAME => application_certificate = Some(v),
+                APPLICATION_CERTIFICATE_THUMBPRINT_NAME => {
+                    application_certificate_thumbprint = Some(v)
+                }
+                AUTHORITY_ID_NAME => authority_id = Some(v),
+                MSI_PARAMS_NAME => msi_params = Some(v),
+                FEDERATED_SECURITY_NAME => match v {
+                    "true" => federated_security = Some(true),
```

Comment:
DRY - you're doing this exact comparison for multiple props

Reply:
moved boolean parsing into its own reusable function.